### ISSUE TITLE
Error: Html alias has already been loaded with v3.2.1

### ISSUE THREAD

**Issue by Holt59** [Holt59/cakephp3-bootstrap-helpers#109]
Issue from #107 

@nrother

---

`dev-master` has currently set CakePHP < 3.4.0 as a requirement. Using v.3.2.1.x-dev throws an exception:
```
The "Html" alias has already been loaded with the following config: array (...) which differs from array ( 'className' => 'Bootstrap.Html', ) 
```

**Comment by Holt59** [comment]
When is this exception thrown? Could you show me the code where you load the helpers?

I have changed the names of the helpers in v3.2.1 (`Bootstrap.BootstrapHtmlHelper` -> `Bootstrap.HtmlHelper`), but the old names are still there for compatibility so this should not be an issue...

I cannot reproduce the error with a fresh CakePHP 3.4.0 install with v3.2.1.

---

After a bit of testing, this seems to happens if you load the `Bootstrap.HtmlHelper` twice, but with different configs, e.g.:

```php
// In your AppController
public $helpers = [
    'Html' => [
        'className' => 'Bootstrap.BootstrapHtml'
    ]
];

// In your AppView
public function initialize() {
    $this->loadHelper('Html', [
        'className' => 'Bootstrap.Html'
    ]);
}
```

I don't think it is directly related to the helpers... Did you modify anything related to the names of the helpers?

**Comment by nrother** [comment]
I did not modify any of the settings, I think it must be releated to one
of your last changes... I'll look into this tomorrow.

I just load the helper once in the AppController, but maybe I'm doing
something strange there...

Am 16.02.2017 um 21:53 schrieb Mikaël Capelle:

**Comment by Holt59** [comment]
Could you enable debug and show me the stack trace?

**Comment by Holt59** [comment]
Could you:

- Show me exactly how you load the helpers? In particular if you changed something to `Bootstrap.Html`?
- Enable debug and show me a stack trace of the error?
- Try to clear the cache and see if the errors persists?

**Comment by nrother** [comment]
This is how I load the helpers in my `AppController`:
```php
public $helpers = [
	'Html' => [
		'className' => 'Bootstrap.BootstrapHtml'
	],
	'Form' => [
		'className' => 'Bootstrap.BootstrapForm',
		'widgets' => [
		    'datePicker' => ['DatePicker', '_view'],
		    'static' => ['StaticControl'],
		],
		'templates' => [
		    'staticControl' => '<p class="form-control-static">{{value}}</p>',
		],
		'columns' => [
		    'label' => 2,
		    'input' => 6,
		    'error' => 4
		]
	]
];
```
Stacktrace:
```
⟩ Cake\View\Helper\FormHelper->postLink
⟩ Cake\View\Helper->__get
⟩ Cake\View\View->loadHelper
⟩ Cake\Core\ObjectRegistry->load
⟩ Cake\Core\ObjectRegistry->_checkDuplicate
```
This line seems to be the problem:
```php
$out .= $this->Html->link($title, $url, $options);
```

The error is throw at `FormHelper::postLink`, which seems to be the first call to the FormHelper which is *not* overridden by the Plugin. The same error is triggered by calling `Paginator::numbers()`, even though this method *is* overridden by the plugin...

The problem is fixed by changing all helper names to the new version (without "Bootstrap").

**Comment by Holt59** [comment]
I see where the problem comes from, actually it also happens if you load the standard `Html` helper together with the Bootstrap ones.

The helpers used internally be other helpers (the `Form` helper here) are synchronized with the helpers used by the view. Since the `Form` helpers is using `Bootstrap.Html` internally, there is a conflict if you use `Bootstrap.BootstrapHtml` or `Html` (the standard Cake helper) for your view.

This is something that also happens with 3.1.0 if you use the standard CakePHP helper, so I will try to fix this.

**Comment by Holt59** [comment]
This should be fixed by f4e5bd66c71a62323974fd9a0ebfaf4b125c0edb.

**Note:** The helpers are now linked to the helpers used by the view, so some restrictions apply:

- the `Html` helper set up for the view must be `Bootstrap.Html` in order to use the easy-icon feature;
- the `searchForm()` and `button()` methods of the navbar helper will not work as expected if the `Form` helper set for the view is not `Bootstrap.Form`.

I will add these notes to the documentation as soon as possible.

**Comment by Holt59** [comment]
I am closing this since this should be fixed by now, feel free to (re-)open if you still have troubles!
